tower-hedge/src/lib.rs

```diff
@@ -0,0 +1,199 @@
+}
+pub struct SelectPolicy<P> {
+    policy: P,
+    histo: Histo,
+    min_data_points: u64,
+}
+
+/// A middleware that pre-emptively retries requests which have been outstanding
+/// for longer than a given latency percentile.  If either of the original
+/// future or the retry future completes, that value is used.
+pub fn service<S, P, Request>(
+    service: S,
+    policy: P,
+    min_data_points: u64,
+    latency_percentile: f32,
+    period: Duration,
+) -> Service<S, P>
+where
+    S: tower_service::Service<Request> + Clone,
+    S::Error: Into<Error>,
+    P: Policy<Request> + Clone,
+{
+    let histo = Arc::new(Mutex::new(RotatingHistogram::new(period)));
+    service_with_histo(service, policy, min_data_points, latency_percentile, histo)
+}
+
+/// A hedge middleware with a prepopulated latency histogram.  This is usedful
+/// for integration tests.
+pub fn service_with_mock_latencies<S, P, Request>(
+    service: S,
+    policy: P,
+    min_data_points: u64,
+    latency_percentile: f32,
+    period: Duration,
+    latencies_ms: &[u64],
+) -> Service<S, P>
+where
+    S: tower_service::Service<Request> + Clone,
+    S::Error: Into<Error>,
+    P: Policy<Request> + Clone,
+{
+    let histo = Arc::new(Mutex::new(RotatingHistogram::new(period)));
+    {
+        let mut locked = histo.lock().unwrap();
+        for latency in latencies_ms.iter() {
+            locked.read().record(*latency).unwrap();
+        }
+    }
+    service_with_histo(service, policy, min_data_points, latency_percentile, histo)
+}
+
+fn service_with_histo<S, P, Request>(
+    service: S,
+    policy: P,
+    min_data_points: u64,
+    latency_percentile: f32,
+    histo: Histo,
+) -> Service<S, P>
+where
+    S: tower_service::Service<Request> + Clone,
+    S::Error: Into<Error>,
+    P: Policy<Request> + Clone,
+{
+    // Clone the underlying service and wrap both copies in a middleware that
+    // records the latencies in a rotating histogram.
+    let recorded_a = Latency::new(histo.clone(), service.clone());
+    let recorded_b = Latency::new(histo.clone(), service);
+
+    // Check policy to see if the hedge request should be issued.
+    let filtered = Filter::new(recorded_b, PolicyPredicate(policy.clone()));
+
+    // Delay the second request by a percentile of the recorded request latency
+    // histogram.
+    let delay_policy = DelayPolicy {
+        histo: histo.clone(),
+        latency_percentile,
+    };
+    let delayed = Delay::new(delay_policy, filtered);
+
+    // If the request is retryable, issue two requests -- the second one delayed
+    // by a latency percentile.  Use the first result to complete.
+    let select_policy = SelectPolicy {
+        policy,
+        histo,
+        min_data_points,
+    };
+    Select::new(select_policy, recorded_a, delayed)
+}
+
+// TODO: Remove when Duration::as_millis() becomes stable.
+fn duration_as_millis(d: Duration) -> u64 {
+    d.as_secs() * 1_000 + d.subsec_millis() as u64
```

Comment:
We may want to use checked arithmetic here, to protect again overflows in production (overflow detection is disabled when `--release`).

Reply:
```rust
    pub fn millis(duration: Duration) -> u64 {
        // Round up.
        let millis = (duration.subsec_nanos() + NANOS_PER_MILLI - 1) / NANOS_PER_MILLI;
        duration.as_secs().saturating_mul(MILLIS_PER_SEC).saturating_add(u64::from(millis))
}
```